lightning-net-tokio/src/lib.rs

```diff
@@ -1,277 +1,375 @@
+					// hasn't received the last send yet, it doesn't matter.
+				},
+				_ => panic!()
 			}
-			Ok(())
-		}));
+		}
 	}
 
-	fn new(event_notify: mpsc::Sender<()>, stream: TcpStream) -> (futures::stream::SplitStream<tokio_codec::Framed<TcpStream, tokio_codec::BytesCodec>>, Arc<Mutex<Self>>) {
-		let (writer, reader) = tokio_codec::Framed::new(stream, tokio_codec::BytesCodec::new()).split();
-		let (send_sink, send_stream) = mpsc::channel(3);
-		tokio::spawn(writer.send_all(send_stream.map_err(|_| -> std::io::Error {
-			unreachable!();
-		})).then(|_| {
-			future::result(Ok(()))
-		}));
-		let us = Arc::new(Mutex::new(Self { writer: Some(send_sink), event_notify, pending_read: Vec::new(), read_blocker: None, read_paused: false, need_disconnect: true, id: ID_COUNTER.fetch_add(1, Ordering::AcqRel) }));
+	fn new(event_notify: mpsc::Sender<()>, stream: TcpStream) -> (io::ReadHalf<TcpStream>, mpsc::Receiver<()>, Arc<Mutex<Self>>) {
+		// We only ever need a channel of depth 1 here: if we returned a non-full write to the
+		// PeerManager, we will eventually get notified that there is room in the socket to write
+		// new bytes, which will generate an event. That event will be popped off the queue before
+		// we call write_event, ensuring that we have room to push a new () if, during the
+		// write_event() call, send_data() returns a non-full write.
+		let (write_avail, receiver) = mpsc::channel(1);
+		let (reader, writer) = io::split(stream);
 
-		(reader, us)
+		(reader, receiver,
+		Arc::new(Mutex::new(Self {
+			writer: Some(writer), event_notify, write_avail,
+			read_blocker: None, read_paused: false, need_disconnect_event: true, disconnect: false,
+			id: ID_COUNTER.fetch_add(1, Ordering::AcqRel)
+		})))
 	}
+}
 
-	/// Process incoming messages and feed outgoing messages on the provided socket generated by
-	/// accepting an incoming connection (by scheduling futures with tokio::spawn).
-	///
-	/// You should poll the Receive end of event_notify and call get_and_clear_pending_events() on
-	/// ChannelManager and ChannelMonitor objects.
-	pub fn setup_inbound<CMH: ChannelMessageHandler + 'static>(peer_manager: Arc<peer_handler::PeerManager<SocketDescriptor<CMH>, Arc<CMH>>>, event_notify: mpsc::Sender<()>, stream: TcpStream) {
-		let (reader, us) = Self::new(event_notify, stream);
+/// Process incoming messages and feed outgoing messages on the provided socket generated by
+/// accepting an incoming connection (by scheduling futures with tokio::spawn).
+///
+/// The returned future will complete when the peer is disconnected and associated handling
+/// futures are freed.
+///
+/// You should poll the Receive end of event_notify and call get_and_clear_pending_events() on
+/// ChannelManager and ChannelMonitor objects after each receive event. In general such an event
+/// should occur after nearly every read event, as any read event may trigger an event which must
+/// be handled.
```

Comment:
I'm a little confused by the uses of "event" in this documentation. Specifically, I'm having a hard time differentiating between "read event", "receive event", and the more generic "event" (e.g., "as any read event may trigger an event"). Is the term "event" being overloaded here to refer to two different things?